File: algorithms/105. Construct Binary Tree from Preorder and Inorder Traversal/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        def dfs(stop):
            if preorder and inorder[-1] != stop:
                root = TreeNode(preorder.pop())
                root.left = dfs(root.val)
                inorder.pop()
                root.right = dfs(stop)
                return root
        preorder, inorder = preorder[::-1], inorder[::-1]
        return dfs(None)

# 202 / 202 test cases passed.
# Status: Accepted
# Runtime: 89 ms
# beats 88.80 %

File: algorithms/105. Construct Binary Tree from Preorder and Inorder Traversal/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        deque<int> pre(preorder.begin(), preorder.end()), in(inorder.begin(), inorder.end());
        return dfs(pre, in, (TreeNode *)NULL);
    }

    TreeNode * dfs(deque<int> & pre, deque<int> & in, TreeNode * stop) {
        if ( in.empty() || (stop && in.front() == stop->val) ) {
            return NULL;
        }
        TreeNode * root = new TreeNode(pre.front());
        pre.pop_front();
        root->left = dfs(pre, in, root);
        in.pop_front();
        root->right = dfs(pre, in, stop);
        return root;
    }
};

// 202 / 202 test cases passed.
// Status: Accepted
// Runtime: 9 ms
// beats 89.09 %
